mixer/pkg/runtime2/handler/safe.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handler
+
+import (
+	"context"
+	"fmt"
+
+	"istio.io/istio/mixer/pkg/adapter"
+)
+
+// safeHandlerBuilder is a wrapper around adapter.Builder and provides panic protection.
+type safeHandlerBuilder struct {
+	b adapter.HandlerBuilder
+}
+
+// SafeHandler is a wrapper around adapter.Handler and provides panic protection.
+type SafeHandler struct {
+	h adapter.Handler
+}
+
+// SetAdapterConfig calls the same-named method on the wrapped builder. Returns error if the call panics.
+func (s safeHandlerBuilder) SetAdapterConfig(c adapter.Config) (err error) {
+	// Try to detect panic, even if panic was called with nil.
+	reachedEnd := false
+	defer func() {
+		if reachedEnd {
+			return
+		}
+
+		r := recover()
```

Comment:
Curious : why do it using reachedEnd as oppose to checking if r=:recover() ; r != nil 

Is it for the case panic(nil). We can prevent that through a linter for adapters( whenever we have them).